[iOS] Safari sometimes crashes with an NSInternalInconsistencyException after long pressing on an image

whsieh opened the pull request:
#### 4d8b0ef1158ffb8caff351c0e4b11cf07f4b7298
<pre>
[iOS] Safari sometimes crashes with an NSInternalInconsistencyException after long pressing on an image
<a href="https://bugs.webkit.org/show_bug.cgi?id=243464">https://bugs.webkit.org/show_bug.cgi?id=243464</a>
rdar://97436180

Reviewed by Chris Dumez.

After the changes in &lt;commits.webkit.org/250218@main&gt;, I started using `WTF::CallbackAggregator` to
allow the context menu to present (or continue presenting) after all three of:

- Visual Look Up analysis
- Object lifting (for &quot;Copy Subject&quot;)
- Machine-readable-code analysis

...have completed. To achieve this, I created a `Ref&lt;CallbackAggregator&gt;` that wraps the overall
`CompletionHandler`, and copied this `Ref` as a captured variable in each of the Objective-C blocks
that are passed into the three asynchronous analysis methods. When the ref-count of this aggregator
reaches 0, we then invoke the overall completion handler. However, there are corner cases in which
VisionKitCore might:

(1) ...never invoke the completion block we pass to it, and also
(2) ...destroy the completion block we give it on a background thread

In this scenario, the last reference to the `CallbackAggregator` may be removed on a background
thread, which means we&apos;ll end up trying to present (or update) the context menu interaction from the
background, thereby triggering this exception.

To fix this, we instead create and deploy an alternate version of `CallbackAggregator` that hops to
the main thread before calling the completion handler by subclassing `ThreadSafeRefCounted` with the
`DestructionThread::MainRunLoop` template parameter, in the case where the destructor is invoked on
the background.

* Source/WTF/wtf/CallbackAggregator.h:
(WTF::CallbackAggregatorOnThread::create):
(WTF::CallbackAggregatorOnThread::~CallbackAggregatorOnThread):
(WTF::CallbackAggregatorOnThread::CallbackAggregatorOnThread):
(WTF::CallbackAggregator::create): Deleted.
(WTF::CallbackAggregator::~CallbackAggregator): Deleted.
(WTF::CallbackAggregator::CallbackAggregator): Deleted.

Add support for `MainRunLoopCallbackAggregator`.

* Source/WebKit/NetworkProcess/cache/CacheStorageEngine.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::removeData):

Replace the existing `RemovalCallbackAggregator` with `MainRunLoopCallbackAggregator` here as well.

(WebKit::RemovalCallbackAggregator::create): Deleted.
(WebKit::RemovalCallbackAggregator::~RemovalCallbackAggregator): Deleted.
(WebKit::RemovalCallbackAggregator::RemovalCallbackAggregator): Deleted.
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _completeImageAnalysisRequestForContextMenu:requestIdentifier:hasTextResults:]):
(-[WKContentView imageAnalysisGestureDidTimeOut:]):

Deploy `MainRunLoopCallbackAggregator` in image analysis code, to ensure that the completion handler
is invoked on the main runloop.

* Source/WebKit/WebProcess/WebCoreSupport/WebBroadcastChannelRegistry.h:

Canonical link: <a href="https://commits.webkit.org/253073@main">https://commits.webkit.org/253073@main</a>
</pre>

cdumez commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -366,6 +365,37 @@ - (void)setImageAnalysis:(CocoaImageAnalysis *)analysis;
+
+    ~ImageAnalysisCallbackAggregator()
+    {
+        if (RunLoop::isMain()) {
```
Could have used `ensureOnMainRunLoop` but I recommend I suggestion above ^ instead.

whsieh replied:
(Yep! Going to go with adding the "call on main thread" variant to WTF)

cdumez commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -366,6 +365,37 @@ - (void)setImageAnalysis:(CocoaImageAnalysis *)analysis;
+    }
+
+private:
+    ImageAnalysisCallbackAggregator(CompletionHandler<void()>&& completionHandler)
```
explicit

cdumez commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -366,6 +365,37 @@ - (void)setImageAnalysis:(CocoaImageAnalysis *)analysis;
 
 #if ENABLE(IMAGE_ANALYSIS)
 
+class ImageAnalysisCallbackAggregator : public ThreadSafeRefCounted<ImageAnalysisCallbackAggregator> {
```
You could just use:
ThreadSafeRefCounted<ImageAnalysisCallbackAggregator, WTF::DestructionThread::MainRunLoop>

Then the destructor only needs to call the completion handler, no need for threading checks.

cdumez replied:
Maybe we could even templatize the WTF one to promote reuse:
```
diff --git a/Source/WTF/wtf/CallbackAggregator.h b/Source/WTF/wtf/CallbackAggregator.h
index 88c5bbd093ba..77451b61611f 100644
--- a/Source/WTF/wtf/CallbackAggregator.h
+++ b/Source/WTF/wtf/CallbackAggregator.h
@@ -32,11 +32,12 @@
 
 namespace WTF {
 
-class CallbackAggregator : public ThreadSafeRefCounted<CallbackAggregator> {
+template <DestructionThread destructionThread>
+class CallbackAggregatorOnThread : public ThreadSafeRefCounted<CallbackAggregator, destructionThread> {
 public:
     static Ref<CallbackAggregator> create(CompletionHandler<void()>&& callback) { return adoptRef(*new CallbackAggregator(WTFMove(callback))); }
 
-    ~CallbackAggregator()
+    ~CallbackAggregatorOnThread()
     {
         ASSERT(m_wasConstructedOnMainThread == isMainThread());
         if (m_callback)
@@ -44,7 +45,7 @@ public:
     }
 
 private:
-    explicit CallbackAggregator(CompletionHandler<void()>&& callback)
+    explicit CallbackAggregatorOnThread(CompletionHandler<void()>&& callback)
         : m_callback(WTFMove(callback))
 #if ASSERT_ENABLED
         , m_wasConstructedOnMainThread(isMainThread())
@@ -58,6 +59,10 @@ private:
 #endif
 };
 
+using CallbackAggregator = CallbackAggregatorOnThread<DestructionThread::Any>;
+using MainRunLoopCallbackAggregator = CallbackAggregatorOnThread<DestructionThread::MainRunLoop>;
+
 } // namespace WTF
 
 using WTF::CallbackAggregator;
+using WTF::MainRunLoopCallbackAggregator;
```

whsieh replied:
Yeah, I considered adding it to `WTF::CallbackAggregator`, but wasn't sure if I should since we're only using it in this file. Sounds like this is okay though, so I'll update this patch to use `MainRunLoopCallbackAggregator`. Thanks!

cdumez replied:
I think this is useful and likely to be reusable. As a matter of fact, I think `RemovalCallbackAggregator` in WebsiteDataStore.cpp might be able to use it.

whsieh replied:
Hm...that helper class already seems to assert that the destructor is only called on the main runloop, though:
```
    ~RemovalCallbackAggregator()
    {
        ASSERT(RunLoop::isMain());
        RunLoop::main().dispatch(WTFMove(m_completionHandler));
    }
```
(Though, maybe there are still cases where it can be destroyed on the background, perhaps via API misuse by the client?)

cdumez replied:
Because it is asking for its destructor to be called on the main run loop via its base class:
`class RemovalCallbackAggregator : public ThreadSafeRefCounted<RemovalCallbackAggregator, WTF::DestructionThread::MainRunLoop> {`

Exactly like your new MainRunLoopCallbackAggregator..

whsieh replied:
~The difference is that `~MainRunLoopCallbackAggregator()` may be called on the background though, whereas  `~RemovalCallbackAggregator()` is intended to be invoked on the main runloop only (the thing they do have in common is that both will invoke their callbacks on the main runloop).~

~I think I would prefer to replace `RemovalCallbackAggregator` with `MainRunLoopCallbackAggregator` in a separate patch, since it's a very subtle behavior change (and isn't necessary to fix this particular crash).~

(Never mind, I see what you mean...)

cdumez commented on Source/WTF/wtf/CallbackAggregator.h:
```diff
@@ -32,19 +32,27 @@
 
-class CallbackAggregator : public ThreadSafeRefCounted<CallbackAggregator> {
+template <DestructionThread destructionThread>
+class CallbackAggregatorOnThread : public ThreadSafeRefCounted<CallbackAggregatorOnThread<destructionThread>> {
```
The idea was to pass destructionThread as template parameter to ThreadSafeRefCounted.

cdumez commented on Source/WTF/wtf/CallbackAggregator.h:
```diff
@@ -32,19 +32,27 @@
-        if (m_callback)
-            m_callback();
+        if constexpr (destructionThread == DestructionThread::MainRunLoop) {
+            ensureOnMainRunLoop([callback = WTFMove(m_callback)]() mutable {
```
This way, we wouldn't need logic like this. The destructor would get called on the thread that the client expected.

whsieh replied:
Yeah, that makes sense. I didn't realize `ThreadSafeRefCounted<DestructionThread::MainRunLoop>` also ensured that the destructor would be invoked on the main runloop, despite the obvious name 🤦🏻‍♂️

cdumez reviewed (approved):
Nice.

whsieh committed:
[iOS] Safari sometimes crashes with an NSInternalInconsistencyException after long pressing on an image
https://bugs.webkit.org/show_bug.cgi?id=243464
rdar://97436180

Reviewed by Chris Dumez.

After the changes in <commits.webkit.org/250218@main>, I started using `WTF::CallbackAggregator` to
allow the context menu to present (or continue presenting) after all three of:

- Visual Look Up analysis
- Object lifting (for "Copy Subject")
- Machine-readable-code analysis

...have completed. To achieve this, I created a `Ref<CallbackAggregator>` that wraps the overall
`CompletionHandler`, and copied this `Ref` as a captured variable in each of the Objective-C blocks
that are passed into the three asynchronous analysis methods. When the ref-count of this aggregator
reaches 0, we then invoke the overall completion handler. However, there are corner cases in which
VisionKitCore might:

(1) ...never invoke the completion block we pass to it, and also
(2) ...destroy the completion block we give it on a background thread

In this scenario, the last reference to the `CallbackAggregator` may be removed on a background
thread, which means we'll end up trying to present (or update) the context menu interaction from the
background, thereby triggering this exception.

To fix this, we instead create and deploy an alternate version of `CallbackAggregator` that hops to
the main thread before calling the completion handler by subclassing `ThreadSafeRefCounted` with the
`DestructionThread::MainRunLoop` template parameter, in the case where the destructor is invoked on
the background.

* Source/WTF/wtf/CallbackAggregator.h:
(WTF::CallbackAggregatorOnThread::create):
(WTF::CallbackAggregatorOnThread::~CallbackAggregatorOnThread):
(WTF::CallbackAggregatorOnThread::CallbackAggregatorOnThread):
(WTF::CallbackAggregator::create): Deleted.
(WTF::CallbackAggregator::~CallbackAggregator): Deleted.
(WTF::CallbackAggregator::CallbackAggregator): Deleted.

Add support for `MainRunLoopCallbackAggregator`.

* Source/WebKit/NetworkProcess/cache/CacheStorageEngine.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::removeData):

Replace the existing `RemovalCallbackAggregator` with `MainRunLoopCallbackAggregator` here as well.

(WebKit::RemovalCallbackAggregator::create): Deleted.
(WebKit::RemovalCallbackAggregator::~RemovalCallbackAggregator): Deleted.
(WebKit::RemovalCallbackAggregator::RemovalCallbackAggregator): Deleted.
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _completeImageAnalysisRequestForContextMenu:requestIdentifier:hasTextResults:]):
(-[WKContentView imageAnalysisGestureDidTimeOut:]):

Deploy `MainRunLoopCallbackAggregator` in image analysis code, to ensure that the completion handler
is invoked on the main runloop.

* Source/WebKit/WebProcess/WebCoreSupport/WebBroadcastChannelRegistry.h:

Canonical link: https://commits.webkit.org/253073@main
